### PR TITLE
Remove overload of GPU max_pool3d with kernel_width; fix nan, inf in GPU {fractional,adaptive} max_pool{2,3}d

### LINKED ISSUE #39044: nn.FractionalMaxPooling on cuda does not propagate nans, and does not work for -inf-filled tensors
## 🐛 Bug

Per title
```
import torch
import math
m = torch.nn.FractionalMaxPool2d(3, output_size=(4, 4))
inp=torch.randn(2, 2, 30, 30, device="cuda")
inp[0][0][0][0] = math.nan
print(m(inp))  # no nan in the input
inp.fill_(-math.inf)
print(m(inp)) # prints tensor filled with finfo.min, would assert if asserts were enabled
```

## Expected behavior

Propagating nans, not asserting/not filling outputs with incorrect values for -inf inputs

cc @ngimel

### LINKED ISSUE #39846: NaN handling in cuda nn.MaxPool3d
## 🐛 Bug

NaN becomes a negative float after nn.MaxPool3d

## To Reproduce

```
import numpy as np
import torch

mp = torch.nn.MaxPool3d(kernel_size=2, stride=2, dilation=1)
empty = torch.tensor([np.nan]*(4**3)).reshape([1, 1, 4, 4, 4]).cuda()
print (mp(empty))

# tensor([[[[[-3.4028e+38, -3.4028e+38],
#            [-3.4028e+38, -3.4028e+38]],

#           [[-3.4028e+38, -3.4028e+38],
#            [-3.4028e+38, -3.4028e+38]]]]], device='cuda:0')
```

## Expected behavior

```
tensor([[[[nan, nan],
          [nan, nan]]]], device='cuda:0')
```

## Environment

PyTorch version: 1.5.0+cu101
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 435.21
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.5.0

Versions of relevant libraries:
[pip] numpy==1.18.2
[pip] torch==1.5.0+cu101
[pip] torchcontrib==0.0.2
[conda] numpy                     1.18.2                   pypi_0    pypi
[conda] torch                     1.5.0+cu101              pypi_0    pypi
[conda] torchcontrib              0.0.2                    pypi_0    pypi

## Additional context

nn.MaxPool2d and cpu version of nn.MaxPool3d works fine.

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by xwang233**
Fix #39846.
Fix #39044

The problem was that `max_pool3d_with_indices_single_out_frame` has an overload of kernel_width being a template argument. The two overloaded kernels were supposed to be identical, however, they were not. 

The general version
https://github.com/pytorch/pytorch/blob/da3073e9b1db503f106842339f50f522d973be84/aten/src/ATen/native/cuda/DilatedMaxPool3d.cu#L69-L73

The overloaded version
https://github.com/pytorch/pytorch/blob/da3073e9b1db503f106842339f50f522d973be84/aten/src/ATen/native/cuda/DilatedMaxPool3d.cu#L130-L134

While the max_pool3d being "switch-case"-ed to the overloaded version, the NaN value comparison is ignored. Also, maintaining two overloaded versions of such a complicated kernel would be hard. I'm not sure if the overloaded version would even give huge performance benefit. So I propose to remove the kernel_width overloaded version.

Also, the current test of max_pool_XD_nan forgot the device kwarg. I added that. 

Edit: profiling before and after
script: https://github.com/xwang233/code-snippet/blob/master/maxpool-3d-kw-template-arg/a.py
plot: https://github.com/xwang233/code-snippet/blob/master/maxpool-3d-kw-template-arg/b.ipynb

The performance difference is within +- 5%.

### CODE REVIEW

**Comment by xwang233** [comment]
cc @ptrblck

**Comment by ezyang** [comment]
@xwang233 it doesn't have to be comprehensive, but could you do a quick before/after performance comparison? Thanks!

**Comment by ngimel** [comment]
Could you also roll the fix to #39044 here? I suspect it's a very similar issue.

**Comment by ngimel** [comment]
Agree with @ezyang, usually template arguments provide a significant perf benefits to kernels like this, so it might make sense to put in a one-line fix to the original kernel, depending on the results of your perf tests.

**Comment by xwang233** [comment]
Thanks for the comments. I did some profilings for max_pool3d before and after this PR. The script and plot are available in the links in the updated original post. The performance difference is less than +- 5%, within fluctuation range. 

From my experience 
```c++
int wEnd = min(wStart + (KERNEL_WIDTH - 1) * dilationW + 1, iwidth);
...
for (int w = wStart; w < wEnd; w += dilationW)
```
`kernel_width` is multiplied by a runtime variable (`dilationW`) and wrapped by a `min` function, which makes compiler optimization much less effective. A much better optimization is that we can write a specialized kernel with all default values `dilation=1`, `padding=0`, `stride=kernel_size`, and that can give a lot performance gain. However, this also increases the maintenance cost.

**Comment by xwang233** [comment]
I also updated the nan value handling in fractional max pool 2d and 3d. Tests are also added. However, CPU fractional maxpool yields an internal error from this test
```
RuntimeError: maxVal != -std::numeric_limits<scalar_t>::infinity() INTERNAL ASSERT FAILED at 
"../aten/src/ATen/native/FractionalMaxPool2d.cpp":83, please report a bug to PyTorch. 
```
So I make the test `onlyCUDA`. I can open another issue for this CPU failure.

**Comment by ngimel** [comment]
Cool, can you also please handle fractional_max_pool issue with -inf-filled tensors, and remove asserts from the kernel (because they are ignored anyway)?

**Comment by ngimel** [comment]
maxval is still set as numeric_limits::lowest, which is wrong if your pooling window has only -inf and/or numeric_limits::lowest.

**Comment by xwang233** [comment]
Thanks! I'm testing on it.

**Comment by xwang233** [comment]
I fixed those `-inf` from `{adaptive, fractional} max pooling` and added tests. The CPU `adaptive_max_pool` does not propagate `-inf`, so I left a TODO there.

File: aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu [hunk 1]
```diff
@@ -77,7 +77,7 @@ __global__ void adaptivemaxpool(T *input, T *output, int64_t *indices,
       T *ptr_output = output + oh*osizeW + ow;
       int64_t *ptr_ind = indices + oh*osizeW + ow;
       int argmax = -1;
-      T max = THCNumerics<T>::min();
+      T max = at::numeric_limits<T>::lower_bound(); // -Infinity
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:80]: If there are only -inf in the pooling window, argmax will not be assigned, and there will be invalid memory access in backward. Your tests don't test backward.

**Review by ngimel — CHANGES_REQUESTED** [review]
Can you also please file an issue for CPU errors?

File: aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu (hunk 1 above)
Comment (xwang233) [reply aten/src/ATen/native/cuda/AdaptiveMaxPooling2d.cu:80]: Fixed. `argmax` is now initialized with a value similar to those in the loop.

**Review by ngimel — APPROVED** [review]

**Comment by xwang233** [comment]
ROCm failure seems real. Do you mind if I add `@skipCUDAIfRocm`?
```
01:31:26 ======================================================================
01:31:26 ERROR: test_fractional_max_pool_nan_inf_cuda_bfloat16 (__main__.TestNNDeviceTypeCUDA)
01:31:26 ----------------------------------------------------------------------
01:31:26 Traceback (most recent call last):
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 769, in wrapper
01:31:26     method(*args, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 241, in instantiated_test
01:31:26     result = test(self, device_arg, dtype)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_device_type.py", line 506, in only_fn
01:31:26     return fn(slf, device, *args, **kwargs)
01:31:26   File "test_nn.py", line 10979, in test_fractional_max_pool_nan_inf
01:31:26     res = fn(x)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/modules/module.py", line 609, in _call_impl
01:31:26     result = self.forward(*input, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/modules/pooling.py", line 750, in forward
01:31:26     _random_samples=self._random_samples)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/_jit_internal.py", line 243, in fn
01:31:26     return if_false(*args, **kwargs)
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/functional.py", line 410, in _fractional_max_pool2d
01:31:26     _random_samples)[0]
01:31:26   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/nn/functional.py", line 395, in fractional_max_pool2d_with_indices
01:31:26     return torch._C._nn.fractional_max_pool2d(input, kernel_size, output_size, _random_samples)
01:31:26 RuntimeError: "fractional_max_pool2d_out_cuda_frame" not implemented for 'BFloat16'
```

**Comment by ngimel** [comment]
Yes, please change dtypesIfCUDA

**Pull request closed without merging**